docs/contract-updatability.md

```diff
@@ -0,0 +1,340 @@
+imposed on changing fields of a contract.
+
+### Nested Declarations
+Contracts can have nested composite type declarations such as structs, resources, interfaces, and enums.
+When a contract is updated, its nested declarations are also being checked, because:
+ - They can be used as type annotation for the fields of the same contract, directly or indirectly.
+ - Any third-party contract can import the types defined in this contract and use them as type annotations.
+ - Hence, changing the type definition is the same as changing the type annotation of such a field (which is also invalid,
+   as in the [Fields](#fields) section).
+
+Changes that can be done to the nested declarations, and the update restrictions are described in following sections:
+ - [Structs, resources and interface](#structs-resources-and-interfaces)
+ - [Enums](#enums)
+ - [Functions](#functions)
+ - [Constructors](#constructors)
+
+## Fields
+A field may belong to a contract, struct, resource, or interface.
+
+#### Valid Changes:
+- Removing a field is valid
+  ```cadence
+  // Existing contract
+
+  pub contract Foo {
+      var a: String
+      var b: Int
+  }
+
+
+  // Updated contract
+
+  pub contract Foo {
+      var a: String
+  }
+  ```
+  - It leaves data for the removed field unused at the storage, as it is no longer accessible.
+  - However, it does not cause any runtime crashes.
+
+- Changing the order of fields is valid.
+  ```cadence
+  // Existing contract
+
+  pub contract Foo {
+      var a: String
+      var b: Int
+  }
+
+
+  // Updated contract
+
+  pub contract Foo {
+      var b: Int
+      var a: String
+  }
+  ```
+
+#### Invalid Changes
+- Adding a new field is not valid.
+  ```cadence
+  // Existing contract
+
+  pub contract Foo {
+      var a: String
+  }
+
+
+  // Updated contract
+
+  pub contract Foo {
+      var a: String
+      var b: Int      // Invalid new field
+  }
+  ```
+    - Existing stored data won't have the new field.
+    - Decoding stored data will result in garbage or missing values for such fields.
+
+- Changing the type of existing field is not valid.
+  ```cadence
+  // Existing contract
+  
+  pub contract Foo {
+      var a: String
+  }
+
+
+  // Updated contract
+
+  pub contract Foo {
+      var a: Int      // Invalid type change
+  }
+  ```
+    - In an already stored contract, the field `a` would have a value of type `String`.
+    - Changing the type of the field `a` to `Int`, would make the runtime read the already stored `String`
+      value as an `Int`, which will result in deserialization errors.
+    - Changing the field type to a subtype/supertype of the existing type is also not valid, as it would also
+      potentially cause issues while decoding/encoding.
+      e.g:
+      - Changing an `Int64` field to `Int8` - Stored value could be number `624` which exceeds the value space
+        for `Int8`.
+
```

Comment:
It would be good to add a section about changing the access modifier for a field

Reply:
That's a good idea! will ad it too.

Reply:
Added in 0e08133